repair/repair.cc

```diff
@@ -1623,3 +1623,96 @@ future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_me
         rlogger.info("bootstrap_with_repair: finished with keyspaces={}", keyspaces);
     });
 }
+
+future<> do_decommission_removenode_with_repair(seastar::sharded<database>& db, locator::token_metadata tm, gms::inet_address leaving_node) {
+    using inet_address = gms::inet_address;
+    return seastar::async([&db, tm = std::move(tm), leaving_node = std::move(leaving_node)] () mutable {
+        auto myip = utils::fb_utilities::get_broadcast_address();
+        auto keyspaces = db.local().get_non_system_keyspaces();
+        bool is_removenode = myip == leaving_node;
+        auto op = is_removenode ? "removenode_with_repair" : "decommission_with_repair";
+        rlogger.info("{}: started with keyspaces={}, leaving_node={}", op, keyspaces, leaving_node);
+        for (auto& keyspace_name : keyspaces) {
+            if (!db.local().has_keyspace(keyspace_name)) {
+                rlogger.info("{}: keyspace={} does not exist any more, ignoring it", op, keyspace_name);
+                continue;
+            }
+            auto& ks = db.local().find_keyspace(keyspace_name);
+            auto& strat = ks.get_replication_strategy();
+            // First get all ranges the leaving node is responsible for
+            dht::token_range_vector ranges = strat.get_ranges(leaving_node);
+            rlogger.info("{}: started with keyspace={}, leaving_node={}, nr_ranges={}", op, keyspace_name, leaving_node, ranges.size());
+            std::unordered_map<dht::token_range, std::vector<inet_address>> current_replica_endpoints;
+            // Find (for each range) all nodes that store replicas for these ranges as well
+            for (auto& r : ranges) {
+                auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
+                auto eps = strat.calculate_natural_endpoints(end_token, tm);
+                current_replica_endpoints.emplace(r, std::move(eps));
+            }
+            auto temp = tm.clone_after_all_left();
+            // leaving_node might or might not be 'leaving'. If it was not leaving (that is, removenode
+            // command was used), it is still present in temp and must be removed.
+            if (temp.is_member(leaving_node)) {
+                temp.remove_endpoint(leaving_node);
+            }
+            std::unordered_map<dht::token_range, repair_neighbors> range_sources;
+            dht::token_range_vector ranges_for_removenode;
+            for (auto&r : ranges) {
+                seastar::thread::maybe_yield();
```

Comment:
I think it would be a bit cleaner to move this maybe_yield inside `sync_data_with_repair()`.
It's called from every operation, meaning that we can simplify the callers, and is called even for empty ranges

Reply:
Please note, we do not call sync_data_with_repair for each range. We call sync_data_with_repair for all the ranges.  The yield under the ranges loop is to prevent the stall if we have tons of ranges.